src/VisualStudio/Core/Def/Implementation/Options/LanguageSettingsPersister.cs

```diff
@@ -40,24 +39,37 @@ internal sealed class LanguageSettingsPersister : ForegroundThreadAffinitizedObj
         /// but the ngen image will exist for the basic map between two reference types, since those are reused.</remarks>
         private readonly IBidirectionalMap<string, Tuple<Guid>> _languageMap;
 
+        private IVsTextManager4 _textManager;
+        private ComEventSink _textManagerEvents2Sink;
+
         /// <remarks>
         /// We make sure this code is from the UI by asking for all serializers on the UI thread in <see cref="HACK_AbstractCreateServicesOnUiThread"/>.
         /// </remarks>
         [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
         public LanguageSettingsPersister(
-            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
-            IGlobalOptionService optionService)
+            IGlobalOptionService optionService, [Import(typeof(SAsyncServiceProvider))] Shell.IAsyncServiceProvider serviceProvider)
             : base(assertIsForeground: true)
         {
-            _textManager = (IVsTextManager4)serviceProvider.GetService(typeof(SVsTextManager));
             _optionService = optionService;
+            _serviceProvider = serviceProvider;
 
             // TODO: make this configurable
             _languageMap = BidirectionalMap<string, Tuple<Guid>>.Empty.Add(LanguageNames.CSharp, Tuple.Create(Guids.CSharpLanguageServiceId))
                                                                .Add(LanguageNames.VisualBasic, Tuple.Create(Guids.VisualBasicLanguageServiceId))
                                                                .Add("TypeScript", Tuple.Create(new Guid("4a0dddb5-7a95-4fbf-97cc-616d07737a77")))
                                                                .Add("F#", Tuple.Create(new Guid("BC6DD5A5-D4D6-4dab-A00D-A51242DBAF1B")))
                                                                .Add("Xaml", Tuple.Create(new Guid("CD53C9A1-6BC2-412B-BE36-CC715ED8DD41")));
+        }
+
+        public async Task InitializeAsync(CancellationToken cancellationToken)
+        {
+            AssertIsForeground();
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            _textManager = (IVsTextManager4)await _serviceProvider.GetServiceAsync(typeof(SVsTextManager)).ConfigureAwait(true);
+            Assumes.Present(_textManager);
 
             foreach (var languageGuid in _languageMap.Values)
```

Comment:
This bit is still the really expensive bit; we want this off the UI thread. Have we reached out to them for better solutions?

Reply:
this is more of to see whether we want this approach. so, for ones, that can't be moved to BG, we will leave things as it is?

...

this particular one I didn't get response, but other native IVs Services, it looks like it will take sometime for all those to be free-threaded.

for those, current benefit we can expect is moving it to Async but keep it on UI thread. but ones call GetService from ctor on UI thread, that won't be possible.